### PR TITLE
test: more storage-related tests

### PR DESCRIPTION
**Pull request by mrc0mmand**
WIP, follow-up to comments from https://github.com/systemd/systemd/pull/20351.

Things to cover:
  * [x] multipath
    * [x] cover failover (mount the MP device(?), check read/write, call `echo offline > /sys/block/sdX/device/state && echo 1 > /sys/block/sdX/device/delete`, check read/write, rinse & repeat)
  * [x] check for broken symlinks
    * [x] coverage for #19946 
  * [ ] LVM
  * [ ] iSCSI
  * [ ] ???

### CODE REVIEW

**Comment by yuwata** [comment]
It may not be simple to add at least in the current test infra, but please consider iSCSI if possible.

**Comment by mrc0mmand** [comment]
@yuwata I'm trying to add coverage for #19946, but with the current approach (see the latest WIP commit) I've only managed to reproduce the issue once (with 1cd4e325693007b3628f1a27297f0ab7114b24b8 reverted). Is the original issue indeed *that* racy or am I doing something wrong?

**Comment by yuwata** [comment]
Not sure, but what happens when `udevadm settle` and the checker in the loop is dropped, but call them after the loop?

**Comment by mrc0mmand** [comment]
r

> Not sure, but what happens when `udevadm settle` and the checker in the loop is dropped, but call them after the loop?

Unfortunately, the outcome is the same.

**Comment by yuwata** [comment]
What happens when all commits in #20603 are reverted?

**Comment by mrc0mmand** [comment]
Still nothing. I tweaked the test to do the settle & check stuff on every 10th iteration and bumped the # of iterations, yet it still passes even with #20603 reverted (`git revert -m 1 b2cdc2c05ee77405fa00f3a330bd19cb755a01d0`).

**Comment by yuwata** [comment]
Have you seen log message `Found node '%s' with non-matching devnum %s, skipping handling.`?

**Comment by mrc0mmand** [comment]
> Have you seen log message `Found node '%s' with non-matching devnum %s, skipping handling.`?

Nope. I guess I'll give `parted` a try, as it was used in the original issue. Maybe `sfdisk` does some background magic which makes the issue less likely? :shrug:

File: test/units/testsuite-64.sh [hunk 1]
```diff
@@ -4,6 +4,18 @@
...
+# Check if all symlinks under /dev/disk/ are valid
+helper_check_device_symlinks() {
+    local link
+
+    while read -r link; do
+        if [[ ! -e "$link" ]]; then
+            echo >&2 "ERROR: symlink '$link' points to '$(readlink -f "$link")' which doesn't exist"
+            return 1
+        fi
+    done < <(find /dev/disk -type l)
```
Comment (yuwata) [inline test/units/testsuite-64.sh]: Is it possible to also test the symlink target is correct?

Comment (mrc0mmand) [reply test/units/testsuite-64.sh]: That crossed my mind as well, but I'm not sure how to determine the correct target. I though about using `udevadm info` on the symlink and then comparing the target to the `$DEVNAME` property. Or is there a better way?

Comment (yuwata) [reply test/units/testsuite-64.sh]: I guess that should work. `udevadm info` contains both symlink and device name.

**Comment by yuwata** [comment]
I noticed that the original report in #19946 say that we get `-EEXIST`. But I cannot find anything we get `-EEXIST`, expect for the case addressed in #20713. Also, the reporter's environment might be outdated (the reporter said he applied necessary commits, but not sure). So, I think it is not necessary to consider the issue seriously.

**Comment by mrc0mmand** [comment]
> I noticed that the original report in #19946 says that a udev worker got `-EEXIST`. But I cannot find any possibilities udev to get `-EEXIST`, expect for the case addressed in #20713. Also, the reporter's environment might be outdated (the reporter said he applied necessary commits, but not sure). So, I think it is not necessary to consider the issue seriously.

Glad to hear that, thanks! I gave `parted` a try on Friday, but I managed to only trigger bugs in both `parted` and the kernel, not in udev at all. So I guess I'll stick with `sfdisk`.

**Comment by mrc0mmand** [comment]
@yuwata would it make sense to merge this PR as it is? I'm actually curious how stable these test are going to be in the *standard runs*. I'll open a new PR for the LVM (and other) stuff.

**Comment by yuwata** [comment]
foacal-amd64 CI failure is unrelated:
```
--- test-journal-flush begin ---
journal_file_copy_entry failed: Bad message
Assertion 'r >= 0' failed at src/libsystemd/sd-journal/test-journal-flush.c:52, function main(). Aborting.
--- test-journal-flush end ---
```

**Review by yuwata — APPROVED** [review]
LGTM.

File: test/units/testsuite-64.sh [hunk 2]
```diff
@@ -140,6 +140,43 @@ EOF
...
+        echo "ERROR: failed to find the test SCSI block device"
+        return 1
+    fi
+
+    cat >"$partscript" <<EOF
+$(printf 'name="test%d", size=2M\n' {1..50})
+EOF
+
+    # Initial partition table
+    sfdisk -q -X gpt "$blockdev" <"$partscript"
```
Comment (dtardon) [inline test/units/testsuite-64.sh:159]: Nitpick: use here string instead of a temp. file?

Comment (mrc0mmand) [reply test/units/testsuite-64.sh:159]: The same file is used in the loop below, so I guess it's better to keep it in a file for readability.

Comment (dtardon) [reply test/units/testsuite-64.sh:159]: The partscript can be saved into a variable and the variable can be used in here string at both places...

Comment (mrc0mmand) [reply test/units/testsuite-64.sh:159]: I see, I'll consider it with the next batch of patches.

**Pull request merged**